Reenable dynamic linking

aaronmondal opened the pull request:
This was disabled due to bugs in LLVM 12. The LLVM commit we are using is much newer and contains the fix (https://reviews.llvm.org/D110261).

The change now lets the host decide whether to link libcxx and friends statically or dynamically.

Fixes an issue where Gentoo won't be able to build carbon without portage flags that enable static libraries for libcxx and libunwind.

aaronmondal committed:
Reenable dynamic linking

This was disabled due to bugs in LLVM 12. The LLVM commit we are using
is much newer and contains the fix (https://reviews.llvm.org/D110261).

The change now lets the host decide whether to link libcxx and
friends statically or dynamically.

Fixes an issue where Gentoo won't be able to build carbon without portage
flags that enable static libraries for libcxx and libunwind.

chandlerc commented:
This doesn't seem to address some of the comments about how dynamic linking creates a significant increase in space required for the output, and compared to using a very fast linker, ultimately slower build performance.

I also suspect that `--dynamic-mode=off` is not the critical issue, but instead the use of `-static-libstdc++`. However, the use of that flag is important to make non-system toolchains work well at all. There are other ways we could try to work around this, but they add substantial complexity.

I also think we'd like to keep our support surface on Linux simple. There are *so* many distributions with so many different installations of different versions of LLVM and libc++ that supporting all of them is very difficult.

My preference is that we keep our build working the same way rather than opening up this flexibility, and continue to ask users to install LLVM from Homebrew if their system install doesn't support this build mode.

aaronmondal commented:
@chandlerc Thanks for taking a look at this!

> This doesn't seem to address some of the comments about how dynamic linking creates a significant increase in space required for the output, and compared to using a very fast linker, ultimately slower build performance.

Is this actually something one would notice in a repo this small? I removed `--dynamic-mode` because I found it confusing and couldn't see actual benefits. The flag doesn't disable static linking, it just encourages "mostly static" (https://bazel.build/reference/be/c-cpp#cc_binary.linkstatic) linking. IMO that gives a false sense of linking behavior. Even if the default behavior is not optimal for a particular OS I think something like that should be handled by Bazel/rules_cc itself and not by a nonstandard reconfiguration in this repo.

I may be very wrong here though and the space implications may be critical. If it helps I could try to actually benchmark the impact of this flag.

> I also suspect that --dynamic-mode=off is not the critical issue, but instead the use of -static-libstdc++.

Yes, the Gentoo build issue is fixed by the `-static-libcxx` and `-static-libgcc` flags. I just added the `--dynamic-mode` change because it fit the theme :sweat_smile: 

> However, the use of that flag is important to make non-system toolchains work well at all. There are other ways we could try to work around this, but they add substantial complexity.

Why/how is this the case? What would be needed to work around the limitations?

> I also think we'd like to keep our support surface on Linux simple. There are so many distributions with so many different installations of different versions of LLVM and libc++ that supporting all of them is very difficult.

AFAIK most linux distros tend to strongly discourage use of static libraries and are moving towards increasingly non-static package distribution. Setting the `-static-*` flags is not a standard practice and I'd expect this to hurt portability rather than improve/simplify it. Telling users to separately install an additional package manager to install a specific toolchain seems like bad UX to me. It also doesn't seem necessary, considering that this does not even affect the entire toolchain - its just libcxx and libunwind.

chandlerc commented:
> > However, the use of that flag is important to make non-system toolchains work well at all. There are other ways we could try to work around this, but they add substantial complexity.
> 
> Why/how is this the case? What would be needed to work around the limitations?

Otherwise, the DSOs need to be in a path searched by the loader or we need to compute the correct RPATHs, etc.

A simpler model is just to link the standard library directly into the binary which makes it hermetic and just use the system libc.

> > I also think we'd like to keep our support surface on Linux simple. There are so many distributions with so many different installations of different versions of LLVM and libc++ that supporting all of them is very difficult.
> 
> AFAIK most linux distros tend to strongly discourage use of static libraries and are moving towards increasingly non-static package distribution. Setting the `-static-*` flags is not a standard practice and I'd expect this to hurt portability rather than improve/simplify it.

It's trade-offs.

Dynamic linking allows sharing the library and in extremely limited cases independent updates. For a distro as a whole, this makes some sense. But it ends up having limited utility for C++ DSOs because of the incredibly complex ABI.

The flip side is that dynamic linking makes execution *much* slower if there are hot edges between DSOs. And it can leave you subject to subtle and confusing ABI-mismatch bugs.

On the whole, I much prefer static linking. And you see this with many LLVM users who work to static link it because they want the efficiency and don't want to be subject to the ABI complexities. Following the same pattern for the C++ standard library costs little and leaves in a more consistent and hermetic state.

> Telling users to separately install an additional package manager to install a specific toolchain seems like bad UX to me. It also doesn't seem necessary, considering that this does not even affect the entire toolchain - its just libcxx and libunwind.

None of this is likely to be relevant for eventual Carbon users. If we ever get there, we'll have simple packages they can install.

Right now, we're only really targeting folks actively *developing* Carbon, and we get a huge benefit from having a simple and consistent host toolchain so we have a known and very predictable set of C++ features, etc. So I think on balance, it's the right call. The bad-UX of installing a custom host C++ toolchain seems balance by the much better UX of having an exact supported C++ feature set and implementation.

aaronmondal commented:
@chandlerc Sorry for the late reply. I didn't get a notification 😳

Thanks for answering my questions. I think I now get a better picture of things. I mean I definitely agree that static linking is *much* easier to handle.

Btw regarding hermetic, feature-fixed C++ toolchains I'll shamelessly plug my [rules_ll](https://github.com/eomii/rules_ll) here, which I built because we had similar issues to what you described. The differences in host toolchains were problematic enough for us to justify writing an alternative to `rules_cc` :smile: Unfortunately it's still too much under construction to for me to shill it as dev environment, but we'll get there eventually :innocent:

I'll close this PR since I think it goes against the current goals of the project. Should these goals change at a later point in time the originally proposed change can always be quickly recreated.
